### PR TITLE
Initial refactoring of edge storage

### PR DESCRIPTION
**Pull request by litghost**
#### Description

This changes edge storage from an allocation array of struct per node to
struct of array for all edge data.

Several algorithms over edges that were previous per node per edge, but
were actually just iteration over edges are now part of rr_node_storage.

This PR is built on top #1084 .

#### Related Issue

#1079 
#1081 
#1084 

#### Motivation and Context

This changes the number of heap allocations on preallocated array (e.g. loading rr graph from file) from `Nnodes` to `3`.

This changes reduces the number of heap allocations during rr graph building from `Nnodes` to `3 * ceil(log2(Nedges) - log2(10*Nodes))`.

When the number of edges is known the max memory usage for edges is now `Nedge * (2*sizeof(int)+sizeof(short))`.  When the number of edges is not known, the max memory usage for edges is now:

```
min(10*Nnode, Nedge) * (2*sizeof(int)+sizeof(short)) + sizeof(int)* (Nedge/2)
```

#### How Has This Been Tested?

- [ ] Travis CI is green
- [ ] Nightly and weekly QoR metrics are acceptable

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

File: vpr/src/route/rr_node_fwd.h [hunk 1]
```diff
@@ -1,21 +1,54 @@
...
+    edge_idx_iterator(value_type init)
+        : value_(init) {}
+    iterator operator++() {
+        value_ += 1;
+        return *this;
+    }
+    iterator operator--() {
+        value_ -= 1;
+        return *this;
+    }
```
Comment (HackerFoo) [inline vpr/src/route/rr_node_fwd.h]: Why not use the corresponding operators (also `!=` below)? They don't exist?

File: vpr/src/route/rr_node_storage.h [hunk 2]
```diff
@@ -119,15 +130,177 @@ class t_rr_node_storage {
...
+        return vtr::make_range(edge_idx_iterator(0), edge_idx_iterator(num_edges(id) - num_non_configurable_edges(id)));
+    }
+    edge_idx_range non_configurable_edges(const RRNodeId& id) const {
+        return vtr::make_range(edge_idx_iterator(num_edges(id) - num_non_configurable_edges(id)), edge_idx_iterator(num_edges(id)));
+    }
+
+    t_edge_size num_edges(const RRNodeId& id) const {
+        auto first_id = first_edge_[id];
+        auto second_id = (&first_edge_[id])[1];
+        return (size_t)second_id - (size_t)first_id;
```
Comment (HackerFoo) [inline vpr/src/route/rr_node_storage.h]: It's dangerous to use `auto` followed by a cast. I'd prefer an explicit type here.

File: vpr/src/route/rr_node_storage.h [hunk 3]
```diff
@@ -119,15 +130,177 @@ class t_rr_node_storage {
...
+    }
+
+    t_edge_size num_configurable_edges(const RRNodeId& id) const;
+    t_edge_size num_non_configurable_edges(const RRNodeId& id) const;
+
+    RREdgeId edge_id(const RRNodeId& id, t_edge_size iedge) const {
+        auto first_edge = first_edge_[id];
```
Comment (HackerFoo) [inline vpr/src/route/rr_node_storage.h]: Please use an explicit type.

File: vpr/src/route/rr_node_storage.h (hunk 2 above)
Comment (litghost) [reply vpr/src/route/rr_node_storage.h]: This isn't actually a cast, it's a `operator size_t`.

File: vpr/src/route/rr_node_fwd.h (hunk 1 above)
Comment (litghost) [reply vpr/src/route/rr_node_fwd.h]: Sorry what?  This comment doesn't make a lot of sense.

File: vpr/src/route/rr_node_storage.h (hunk 3 above)
Comment (litghost) [reply vpr/src/route/rr_node_storage.h]: Done

**Comment by HackerFoo** [comment]
There are many uses of `first` and `second` seem to be used as placeholder names, like `x` and `y`.

**Comment by litghost** [comment]
Sanity check with `gsm_switch_stratixiv_arch_timing.blif` (from weekly titan tests)

Before:
```
## Build routing resource graph took 212.56 seconds (max_rss 7950.6 MiB, delta_rss +1698.9 MiB)
  RR Graph Nodes: 20014774
  RR Graph Edges: 159049032
# Create Device took 214.84 seconds (max_rss 7950.6 MiB, delta_rss +1698.9 MiB)
```

After:
```
## Build routing resource graph took 159.88 seconds (max_rss 9165.3 MiB, delta_rss +2842.8 MiB)
  RR Graph Nodes: 20014774
  RR Graph Edges: 159049032
# Create Device took 161.22 seconds (max_rss 9165.3 MiB, delta_rss +2842.8 MiB)
```

This jump in ~1.2 GiB is basically the cost of storing the source rr node explicitly.  I'll noddle over alternative strategies.

**Comment by litghost** [comment]
> Sanity check with `gsm_switch_stratixiv_arch_timing.blif` (from weekly titan tests)
> 
> Before:
> 
> ```
> ## Build routing resource graph took 212.56 seconds (max_rss 7950.6 MiB, delta_rss +1698.9 MiB)
>   RR Graph Nodes: 20014774
>   RR Graph Edges: 159049032
> # Create Device took 214.84 seconds (max_rss 7950.6 MiB, delta_rss +1698.9 MiB)
> ```
> 
> After:
> 
> ```
> ## Build routing resource graph took 159.88 seconds (max_rss 9165.3 MiB, delta_rss +2842.8 MiB)
>   RR Graph Nodes: 20014774
>   RR Graph Edges: 159049032
> # Create Device took 161.22 seconds (max_rss 9165.3 MiB, delta_rss +2842.8 MiB)
> ```
> 
> This jump in ~1.2 GiB is basically the cost of storing the source rr node explicitly. I'll noddle over alternative strategies.

So in an unexpected turn, while the max_rss after create device **is higher** than before, because the heap as not threshed, the final max_rss is much better:

Before:
```
The entire flow of VPR took 13322.02 seconds (max_rss 16152.4 MiB)
```

After:
```
The entire flow of VPR took 7068.79 seconds (max_rss 10713.4 MiB)
```

I'll need to do a full `vtr_reg_weekly` to confirm the results.

File: vpr/src/route/rr_node_storage.h [hunk 4]
```diff
@@ -119,15 +130,177 @@ class t_rr_node_storage {
...
+    void assign_edges();
+
+    // Verify that first_edge_ array correctly partitions rr edge data.
+    bool verify_first_edges() const;
+
     vtr::vector<RRNodeId, t_rr_node_data> storage_;
+    vtr::vector<RRNodeId, RREdgeId> first_edge_;
+    vtr::vector<RRNodeId, t_edge_size> fan_in_;
+
+    vtr::vector<RREdgeId, RRNodeId> edge_src_node_;
```
[the remark this answers is not in the document]
Comment (kmurray) [reply vpr/src/route/rr_node_storage.h]: Thinking about this a bit more, given the sorted structure of the edges and that we know the start/end of each node's out-going edges, it seems like we should be able to get the source node for a given edge ID in log(num_nodes) time via binary search against the (sorted) `t_rr_node_data::first_edge_`s.

File: vpr/src/route/rr_node_storage.cpp [hunk 5]
```diff
@@ -0,0 +1,457 @@
...
+ * edge_swapper is a reference for the src/dest/switch tuple, and can convert
+ * to and from t_rr_edge_info, the value_type for edge_sort_iterator.
+ *
+ * edge_compare_src_node_and_configurable_first is a comparision operator
+ * that first partitions the edge data by source rr node, and then by
+ * configurable switches first.  Sorting by this comparision operator means that
+ * the edge data is directly usable for each node by simply slicing the arrays.
+ *
+ * */
+struct edge_swapper {
```
Comment (kmurray) [inline vpr/src/route/rr_node_storage.cpp]: This is a nice way to keep the swapping clean in Struct-of-Arrays form!

File: vpr/src/route/rr_node_storage.h [hunk 6]
```diff
@@ -119,15 +130,177 @@ class t_rr_node_storage {
...
+
+    t_edge_size num_configurable_edges(const RRNodeId& id) const;
+    t_edge_size num_non_configurable_edges(const RRNodeId& id) const;
+
+    RREdgeId edge_id(const RRNodeId& id, t_edge_size iedge) const {
+        RREdgeId first_edge = first_edge_[id];
+        RREdgeId ret((size_t)first_edge + iedge);
+        VTR_ASSERT_SAFE(ret < (&first_edge_[id])[1]);
+        return ret;
+    }
```
Comment (kmurray) [inline vpr/src/route/rr_node_storage.h]: That we can do this easily given the edge ordering is a really nice feature, which I expect will make porting/refactoring downstream code much easier.

File: vpr/src/route/rr_node_storage.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (litghost) [reply vpr/src/route/rr_node_storage.h]: The primary reason for the `edge_src_node_` data is to sort the underlying data.  After that is complete, we can discard the array.

File: vpr/src/route/rr_node_storage.h [hunk 7]
```diff
@@ -119,15 +130,177 @@ class t_rr_node_storage {
...
+    RRNodeId edge_sink_node(const RRNodeId& id, t_edge_size iedge) const {
+        return edge_sink_node(edge_id(id, iedge));
+    }
+    short edge_switch(const RREdgeId& edge) const {
+        return edge_switch_[edge];
+    }
+    short edge_switch(const RRNodeId& id, t_edge_size iedge) const {
+        return edge_switch(edge_id(id, iedge));
+    }
```
[the remark this answers is not in the document]
Comment (litghost) [reply vpr/src/route/rr_node_storage.h]: No doubt.  The primary motivation behind this path is something that is more incremental.

File: vpr/src/route/rr_node_storage.cpp [hunk 8]
```diff
@@ -0,0 +1,457 @@
...
+    }
+
+    edges_read_ = true;
+    VTR_ASSERT(remapped_edges_);
+
+    const auto& device_ctx = g_vpr_ctx.device();
+    std::stable_sort(
+        edge_sort_iterator(this, 0),
+        edge_sort_iterator(this, edge_src_node_.size()),
+        edge_compare_src_node_and_configurable_first(device_ctx.rr_switch_inf));
```
[the remark this answers is not in the document]
Comment (litghost) [reply vpr/src/route/rr_node_storage.cpp]: Your understanding is correct.

**Comment by litghost** [comment]
I think I've identified a way to regain some lost CPU performance.  I've started a vtr_reg_weekly QoR run, but it won't return results for a day or two.

I'll push a new branch today with a rebase onto of master, along with fixes to review comments.

File: vpr/src/route/rr_node_storage.h [hunk 9]
```diff
@@ -119,15 +130,177 @@ class t_rr_node_storage {
...
     t_rr_node_data& get(const RRNodeId& id) {
         return storage_[id];
     }
     const t_rr_node_data& get(const RRNodeId& id) const {
         return storage_[id];
     }
 
+    // Take allocated edges in edge_src_node_/ edge_dest_node_ / edge_switch_
+    // sort, and assign the first edge for each
+    void assign_edges();
```
[the remark this answers is not in the document]
Comment (litghost) [reply vpr/src/route/rr_node_storage.h]: Done

File: vpr/src/route/rr_node_storage.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (litghost) [reply vpr/src/route/rr_node_storage.cpp]: Comment added

File: vpr/src/route/rr_node_storage.cpp [hunk 10]
```diff
@@ -0,0 +1,457 @@
...
+    edges_read_ = true;
+    remapped_edges_ = true;
+}
+
+void t_rr_node_storage::partition_edges() {
+    if (partitioned_) {
+        return;
+    }
+
+    edges_read_ = true;
```
[the remark this answers is not in the document]
Comment (litghost) [reply vpr/src/route/rr_node_storage.cpp]: Comments added

File: vpr/src/route/rr_node_storage.h [hunk 11]
```diff
@@ -71,17 +59,25 @@ struct t_rr_node_data {
         int16_t track_num;
         int16_t class_num;
     } ptc_;
-    t_edge_size fan_in_ = 0;
+
     uint16_t capacity_ = 0;
 };
 
 // RR node and edge storage class.
 class t_rr_node_storage {
```
[the remark this answers is not in the document]
Comment (litghost) [reply vpr/src/route/rr_node_storage.h]: Renamed

**Comment by litghost** [comment]
I believe all feedback has been addressed.  I've kicked off a QoR run, results will be ready in a couple days.  Preliminary results show the CPU cost is gone, with the 4% memory increase remaining.  #1096 resolves the memory increase, and changes it to a 35-60% memory reduction.

**Pull request closed without merging**